website/content/en/docs/best-practices/multi-tenancy.md

```diff
@@ -0,0 +1,101 @@
+weight: 1
+description: This guide describes the best practices concepts to write Operators for Multi-Tenancy solutions.
+---
+
+## Operator Behavioral Guidelines
+
+### NetworkPolicy 
+
+If your Operator create or manage [NetworkPolicy][k8s-network-policy] configurations ensure that your solution:
+
+* apply fine-grained network policies to the extent that is required for your managed application to function properly
+* apply fine-grained network policies to enable your managed application internal components to communicate among each other
+* allow users to disable network policy management (your operator doesn't deploy [NetworkPolicy][k8s-network-policy] instances at all)
+* do not create "allow traffic from everywhere in the cluster" type policies
+
+Usage of [NetworkPolicy][k8s-network-policy] is popular in multi-tenant cluster to provide an extra layer
+of segregation among tenants within SDN solutions. Users typically customize this extensively with goal to disallow network traffic among unrelated
+tenants. Operators that deploy a `accept all traffic from anywhere in the cluster` style policy, are creating an obstacle in
+pursuing this goal, especially if it cannot be disabled. In security-conscious environments policies like these are not allowed
+in production. In such cases your operator should minimally **have the option to prevent [NetworkPolicy][k8s-network-policy] objects from being
+created** and leave these to the user. In more advanced cases, your operator should create [NetworkPolicy][k8s-network-policy] configurations that
+follow the least-privilege principle as well implement deny-all-allow-some rules.
+
+### Resource Requests
+
+Create resource requests for CPU and memory:
+
+* for your Operator's manager `Deployment` (See `config/manager/manager.yaml`)
+* for any `Pods` and `Deployments` which are creates by your Operator
+
+```yml
+  ...
+  # TODO(user): Configure the resources accordingly based on the project requirements.
+  # More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+  resources:
+    limits:
+      cpu: 500m
+      memory: 128Mi
+    requests:
+      cpu: 10m
+      memory: 64Mi
+    ... 
+```
+
+In order to prevent tenants from consuming all the resources on a cluster or blocking other workloads 
+from being scheduled, many production clusters will define [ResourceQuota][k8s-resources-quotas] configuration.
+[ResourceQuota][k8s-resources-quotas] configuration in a namespace will cause all pods without 
+[Requests and limits][k8s-requests-limits] for CPU and memory to be rejected.
+
+Because OLM-managed Operators do not run on Cluster infra nodes, they will consume paid CPU cores. 
+Users will naturally want to limit the maximum amount of resources these non-workload components can take up. 
+For this reason, if you are integrating with [OLM][olm], you should, **always ship your operator `Deployment` with resource requests limits for CPU and memory**.
```

Comment:
I don't think that enforcing CPU limit is a good practice. CPU is compressible and extra cycles (available on a node and not requested) are allocated proportionally to the requests made. As such operator pods would not "eat workload lunch" if the worload pods have set requests according to their appetite. But if there are free cycles why not let operators use them rather than getting them lost?
In regard to quotas, the cluster-admins have always the possibility to set defaults if they still want to enforce limits: https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/cpu-default-namespace/

Reply:
This doc is about the set the requests not the limits. 
However, setting the limits are also a good practice. WHY?

> to a bad configuration change that causes a program to go out of control and use 100% of the available CPU 
Ref:https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-resource-requests-and-limits

Also, because of security reasons as well. See: https://cheatsheetseries.owasp.org/cheatsheets/Docker_Security_Cheat_Sheet.html#rule-7-limit-resources-memory-cpu-file-descriptors-processes-restarts

Set the limits for memory and CPU mainly will ensure that the pod will terminate and be re-deployed if something goes wrong and will avoid DoS/malicious attacks.

Reply:
Please, check that the doc is only speaking about the requests anyway.

Reply:
Hi @dmesser @fgiloux 

This doc is about the set the requests not the limits. 
However, by looking at that I understand that: 

If not set, bad configurations or code implementations can consume all available resources which could affect other components on the clusters. Also, for security reasons such as dos attacks which can take advantage of that. More info: https://cheatsheetseries.owasp.org/cheatsheets/Docker_Security_Cheat_Sheet.html#rule-7-limit-resources-memory-cp[…]le-descriptors-processes-restarts and https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-resource-requests-and-limits

- **For memory:** If exceeds its memory limit, it might be terminated and k8s will restart it.
- **For CPU:** Similar idea, however, it will not be killed for excessive CPU usage. According to the doc[2], the CPU is considered a “compressible” resource and the Operator starts hitting the CPU limits, Kubernetes starts throttling the container. This means the CPU will be artificially restricted, giving a potentially worse performance only. 

For the purpose of this PR, it is fixed and is only speaking about requests.

Reply:
that was changed to explain when happens and added as optional.